TechnicalDocuments/Proposals/Fish_Operator.md

```diff
@@ -0,0 +1,124 @@
+# Introduce "fish" operator
+
+* Author: Ilya Puchka
+* Review Manager:TBD
+
+## Introduction
+
+This proposal it to suggest to use new operator that will allow us to compose functions returning optional values, which is not currently possible.
+
+## Motivation
+
+Since introduction of `>>>` operator it was widely adopted throughout the code base and became an important tool in our tool set. But there are some cases where it's not enough. 
+
+One of these cases is when first item returns optional, which makes us to jump through the hoops:
+
+```swift
+didTap: helpText.map { ^Action.didTapWhatDoesThisMean($0) >>> observer }
+```
+
+## Proposed solution
+
+To handle such case [Prelude](https://github.com/pointfreeco/swift-prelude) defines "fish" operator (with a help of free curried `flatMap` function):
+
+```swift
+infix operator >=>: ForwardComposition
+
+public func >=> <A, B, C>(lhs: @escaping (A) -> B?, rhs: @escaping (B) -> C?) -> (A) -> C? {
+  return lhs >>> flatMap(rhs)
+}
+
+public func flatMap<A, B>(_ a2b: @escaping (A) -> B?) -> (A?) -> B? {
+  return { a in
+    a.flatMap(a2b)
+  }
+}
+```
+
+With this we can convert example code above to much simpler form:
+
+```swift
+didTap: ^helpText >=> Action.didTapWhatDoesThisMean >=> observer
```

Comment:
Why is there a `^` operator?

Reply:
Because we can compose functions only, `^` lifts value `T` to function `() -> T`. It is similar to `const`, but `const` will create `(()) -> T` in this case, which is not the same as `() -> T`

Reply:
To lift the value into a function returning the value, so it can be composed with other functions.